src/interpreter/internal.rs

```diff
@@ -166,10 +166,27 @@ impl<S> Interpreter<S> {
         }?;
 
         // update the local copy of the output
-        let mut output = Output::variable(owner_to_set, amount_to_set, asset_id_to_update);
+        let output = Output::variable(owner_to_set, amount_to_set, asset_id_to_update);
+        self.set_output(out_idx, output)
+    }
+
+    /// update an output both in serialized form in-memory and on the referenced tx
+    pub(crate) fn set_output(&mut self, out_idx: usize, mut output: Output) -> Result<(), RuntimeError> {
+        // verify output type matches out_idx type
+        // Use halt errors as these should already have been validated
+        let tx_out = self.tx.outputs().get(out_idx).ok_or(RuntimeError::Halt(io::Error::new(
+            ErrorKind::Other,
+            "Invalid output index",
+        )))?;
+        if std::mem::discriminant(tx_out) != std::mem::discriminant(&output) {
```

Comment:
nit: We don't use fully qualified names elsewhere in the code. In fact, This would make an eventual `no-std` port harder in the future because the imports are not centralized in the header of the file.
```suggestion
        if mem::discriminant(tx_out) != mem::discriminant(&output) {
```

Reply:
there's going to be a lot of work to port to no-std anyways, since the VM relies so heavily on std::io in many places.